### DISCUSSION TITLE
why-is-string-gethashcode-different-each-time-i-run-my-program-in-net-core/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#78] [Announcements]
# why-is-string-gethashcode-different-each-time-i-run-my-program-in-net-core/

In this post I discuss string.GetHashCode(), why it's randomised, and hash-attacks. I also provide a deterministic implementation you can use when it's safe.

https://andrewlock.net/why-is-string-gethashcode-different-each-time-i-run-my-program-in-net-core/

**Comment by cking27** [comment]
Great article, thank you.  I learned something about hashcode and made use of your GetDeterministicHashCode method.

**Comment by koelvin** [comment]
I can confirm it generate a consistent hash across X86 and x64 and across .NET Framework/Core.
Cheers

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 12 2018, 10:02, in reply to [Joshua Robinson](https://disqus.com/by/disqus_anPIwpZQ2t/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/78#discussioncomment-6054212)</em>

---

<p>Glad you enjoyed the article :) The main reason was size - I needed a 32bit hash rather than the 128+ bits those give.</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 12 2018, 10:00, in reply to [bruce davidson](https://disqus.com/by/brucedavidson/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/78#discussioncomment-6054216)</em>

---

<p>Interesting, that definitely isn't what I expect - I expected the has value to be pseudo random, but expected the same input to give the same output!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 12 2018, 09:59, in reply to [drhaas](https://disqus.com/by/xda-5ed41372123c314b5ce17321da52bc5c/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/78#discussioncomment-6054218)</em>

---

<p>Thanks, glad you like it. If you double check the code snippet I am accessing the value using str[i] and str[i+1].</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at January 13 2019, 05:06, in reply to [scn00zr](https://disqus.com/by/scn00zr/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/78#discussioncomment-6054225)</em>

---

<p>The <code>(hash &lt;&lt; 5) + hash + (hash &gt;&gt; 27)</code> version was only used for 32-bit machines, the <code>(hash &lt;&lt; 5) + hash </code> version is used for 64-bit machines. <a href="https://referencesource.microsoft.com/#mscorlib/system/string.cs,874" rel="nofollow noopener" title="https://referencesource.microsoft.com/#mscorlib/system/string.cs,874">https://referencesource.microsoft.com/#mscorlib/system/string.cs,874</a></p>

[a comment by an automated account is not shown]

**Comment by EricOuellet2** [comment]
Thanks! (Same as cking27) ;-) ... Great article, thank you. I learned something about hashcode and made use of your GetDeterministicHashCode method.

**Comment by mikepharesjr** [comment]
Thank you!
Great article!
Are you aware the GetDeterministicHashCode has a different output when I ran on these two systems;
- Intel64 Family 6 Model 140 Stepping 1 GenuineIntel ~2995 Mhz
- Intel64 Family 6 Model 158 Stepping 13 GenuineIntel ~3696 Mhz
